File: backend/app/services/ai_agents.py
"""
TradoVerse 7-Agent AI Consensus System

This module implements the multi-agent AI system for market analysis:
1. Technical Analysis Agent - Chart patterns, indicators, support/resistance
2. Fundamental Analysis Agent - Financial statements, valuations, earnings
3. Sentiment Analysis Agent - News, social media, market sentiment
4. Risk Management Agent - Position sizing, stop-loss, portfolio risk
5. Market Microstructure Agent - Order flow, liquidity, market depth
6. Macroeconomic Agent - Economic indicators, central bank policies
7. Quantitative Agent - Statistical models, factor analysis
"""
import json
import asyncio
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum
import httpx

from ..core.config import get_settings
from ..schemas.schemas import (
    AgentType, Recommendation, AgentAnalysisResult, ConsensusResponse
)

settings = get_settings()


class AIAgent:
    """Base class for AI agents."""
    
    def __init__(self, agent_type: AgentType):
        self.agent_type = agent_type
        self.name = self._get_agent_name()
        self.system_prompt = self._get_system_prompt()
    
    def _get_agent_name(self) -> str:
        """Get human-readable agent name."""
        names = {
            AgentType.TECHNICAL: "Technical Analysis Agent",
            AgentType.FUNDAMENTAL: "Fundamental Analysis Agent",
            AgentType.SENTIMENT: "Sentiment Analysis Agent",
            AgentType.RISK: "Risk Management Agent",
            AgentType.MICROSTRUCTURE: "Market Microstructure Agent",
            AgentType.MACRO: "Macroeconomic Agent",
            AgentType.QUANT: "Quantitative Agent",
        }
        return names.get(self.agent_type, "Unknown Agent")
    
    def _get_system_prompt(self) -> str:
        """Get the system prompt for this agent type."""
        prompts = {
            AgentType.TECHNICAL: """You are an expert Technical Analysis Agent for TradoVerse.
Your role is to analyze price charts, patterns, and technical indicators to provide trading recommendations.

Focus on:
- Price trends and momentum (moving averages, MACD, RSI)
- Chart patterns (head and shoulders, triangles, flags)
- Support and resistance levels
- Volume analysis
- Fibonacci retracements

Provide your analysis in a structured format with clear reasoning.""",

            AgentType.FUNDAMENTAL: """You are an expert Fundamental Analysis Agent for TradoVerse.
Your role is to analyze company financials, valuations, and business fundamentals.

Focus on:
- Revenue growth and profitability
- P/E ratio, P/B ratio, PEG ratio
- Debt levels and cash flow
- Competitive advantages (moat)
- Management quality and insider activity

Provide your analysis in a structured format with clear reasoning.""",

            AgentType.SENTIMENT: """You are an expert Sentiment Analysis Agent for TradoVerse.
Your role is to analyze market sentiment from news, social media, and investor behavior.

Focus on:
- News sentiment and headlines
- Social media trends and discussions
- Analyst ratings and price targets
- Institutional investor activity
- Fear and greed indicators

Provide your analysis in a structured format with clear reasoning.""",

            AgentType.RISK: """You are an expert Risk Management Agent for TradoVerse.
Your role is to assess and manage trading risks.

Focus on:
- Position sizing recommendations
- Stop-loss and take-profit levels
- Portfolio diversification
- Volatility assessment (VIX, historical volatility)
- Risk/reward ratios
- Maximum drawdown analysis

You have VETO POWER if risk is too high. Provide your analysis in a structured format.""",

            AgentType.MICROSTRUCTURE: """You are an expert Market Microstructure Agent for TradoVerse.
Your role is to analyze order flow, liquidity, and market mechanics.

Focus on:
- Bid-ask spreads
- Order book depth
- Trading volume patterns
- Dark pool activity
- Market maker behavior
- Price impact analysis

Provide your analysis in a structured format with clear reasoning.""",

            AgentType.MACRO: """You are an expert Macroeconomic Agent for TradoVerse.
Your role is to analyze macroeconomic factors affecting markets.

Focus on:
- Interest rates and Fed policy
- Inflation trends (CPI, PPI)
- GDP growth and employment
- Global economic conditions
- Sector rotation patterns
- Currency movements

Provide your analysis in a structured format with clear reasoning.""",

            AgentType.QUANT: """You are an expert Quantitative Agent for TradoVerse.
Your role is to apply statistical and mathematical models to trading.

Focus on:
- Statistical arbitrage opportunities
- Factor analysis (momentum, value, quality)
- Mean reversion signals
- Correlation analysis
- Monte Carlo simulations
- Machine learning predictions

Provide your analysis in a structured format with clear reasoning.""",
        }
        return prompts.get(self.agent_type, "You are a trading analysis agent.")
    
    async def analyze(
        self, 
        symbol: str, 
        market_data: Dict[str, Any]
    ) -> AgentAnalysisResult:
        """Analyze a symbol and return recommendation."""
        
        # Build the analysis prompt
        user_prompt = f"""Analyze {symbol} and provide your trading recommendation.

Market Data:
{json.dumps(market_data, indent=2)}

Respond in the following JSON format:
{{
    "recommendation": "buy" | "sell" | "hold",
    "confidence": 0.0 to 1.0,
    "reasoning": "Your detailed analysis and reasoning",
    "key_factors": ["factor1", "factor2", ...],
    "data_points": {{
        "metric1": value1,
        "metric2": value2
    }}
}}"""

        try:
            # Call LLM API
            response = await self._call_llm(user_prompt)
            
            # Parse response
            result = self._parse_response(response)
            
            return AgentAnalysisResult(
                agent_type=self.agent_type,
                recommendation=result.get("recommendation", Recommendation.HOLD),
                confidence=result.get("confidence", 0.5),
                reasoning=result.get("reasoning", "Analysis completed."),
                data_points=result.get("data_points", {})
            )
        except Exception as e:
            # Return neutral recommendation on error
            return AgentAnalysisResult(
                agent_type=self.agent_type,
                recommendation=Recommendation.HOLD,
                confidence=0.3,
                reasoning=f"Analysis error: {str(e)}",
                data_points={}
            )
    
    async def _call_llm(self, prompt: str) -> str:
        """Call the LLM API."""
        async with httpx.AsyncClient() as client:
            # Use built-in Forge API if available, otherwise use OpenAI-compatible API
            if settings.FORGE_API_URL and settings.FORGE_API_KEY:
                response = await client.post(
                    f"{settings.FORGE_API_URL}/llm/chat/completions",
                    headers={
                        "Authorization": f"Bearer {settings.FORGE_API_KEY}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "messages": [
                            {"role": "system", "content": self.system_prompt},
                            {"role": "user", "content": prompt}
                        ],
                        "temperature": 0.7,
                        "max_tokens": 1000
                    },
                    timeout=30.0
                )
            else:
                # Fallback to OpenAI-compatible API
                api_url = settings.LLM_BASE_URL or "https://api.openai.com/v1"
                api_key = settings.OPENAI_API_KEY or settings.DEEPSEEK_API_KEY
                
                response = await client.post(
                    f"{api_url}/chat/completions",
                    headers={
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": settings.LLM_MODEL,
                        "messages": [
                            {"role": "system", "content": self.system_prompt},
                            {"role": "user", "content": prompt}
                        ],
                        "temperature": 0.7,
                        "max_tokens": 1000
                    },
                    timeout=30.0
                )
            
            response.raise_for_status()
            data = response.json()
            return data["choices"][0]["message"]["content"]
    
    def _parse_response(self, response: str) -> Dict[str, Any]:
        """Parse LLM response into structured data."""
        try:
            # Try to extract JSON from response
            import re
            json_match = re.search(r'\{[\s\S]*\}', response)
            if json_match:
                result = json.loads(json_match.group())
                
                # Normalize recommendation
                rec = result.get("recommendation", "hold").lower()
                if rec in ["buy", "strong_buy", "bullish"]:
                    result["recommendation"] = Recommendation.BUY
                elif rec in ["sell", "strong_sell", "bearish"]:
                    result["recommendation"] = Recommendation.SELL
                else:
                    result["recommendation"] = Recommendation.HOLD
                
                # Ensure confidence is in range
                conf = float(result.get("confidence", 0.5))
                result["confidence"] = max(0.0, min(1.0, conf))
                
                return result
        except (json.JSONDecodeError, ValueError):
            pass
        
        # Fallback parsing
        response_lower = response.lower()
        if "buy" in response_lower and "sell" not in response_lower:
            return {"recommendation": Recommendation.BUY, "confidence": 0.6, "reasoning": response}
        elif "sell" in response_lower and "buy" not in response_lower:
            return {"recommendation": Recommendation.SELL, "confidence": 0.6, "reasoning": response}
        else:
            return {"recommendation": Recommendation.HOLD, "confidence": 0.5, "reasoning": response}


class ConsensusEngine:
    """
    Multi-agent consensus engine that coordinates all 7 agents
    and produces a final trading recommendation.
    """
    
    def __init__(self):
        self.agents: Dict[AgentType, AIAgent] = {
            agent_type: AIAgent(agent_type) for agent_type in AgentType
        }
    
    async def get_consensus(
        self,
        symbol: str,
        market_data: Dict[str, Any],
        enabled_agents: List[AgentType] = None
    ) -> ConsensusResponse:
        """
        Run all enabled agents and compute consensus recommendation.
        
        Args:
            symbol: Stock symbol to analyze
            market_data: Market data for analysis
            enabled_agents: List of agents to use (defaults to all)
        
        Returns:
            ConsensusResponse with final recommendation and agent analyses
        """
        if enabled_agents is None:
            enabled_agents = list(AgentType)
        
        # Run all agents in parallel
        tasks = [
            self.agents[agent_type].analyze(symbol, market_data)
            for agent_type in enabled_agents
            if agent_type in self.agents
        ]
        
        analyses = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Filter out exceptions
        valid_analyses: List[AgentAnalysisResult] = [
            a for a in analyses if isinstance(a, AgentAnalysisResult)
        ]
        
        # Calculate vote breakdown
        vote_breakdown = {"buy": 0, "sell": 0, "hold": 0}
        weighted_scores = {"buy": 0.0, "sell": 0.0, "hold": 0.0}
        
        for analysis in valid_analyses:
            rec = analysis.recommendation.value
            vote_breakdown[rec] += 1
            weighted_scores[rec] += analysis.confidence
        
        # Check for Risk Agent veto
        risk_analysis = next(
            (a for a in valid_analyses if a.agent_type == AgentType.RISK),
            None
        )
        risk_veto = False
        if risk_analysis and risk_analysis.confidence < 0.3:
            risk_veto = True
        
        # Determine final recommendation
        if risk_veto:
            final_recommendation = Recommendation.HOLD
            reasoning = "Risk Management Agent has vetoed the trade due to high risk."
        else:
            # Use weighted voting
            max_score = max(weighted_scores.values())
            final_recommendation = Recommendation(
                max(weighted_scores, key=weighted_scores.get)
            )
            reasoning = self._generate_reasoning_summary(valid_analyses, vote_breakdown)
        
        # Calculate consensus confidence
        total_votes = sum(vote_breakdown.values())
        if total_votes > 0:
            winning_votes = vote_breakdown[final_recommendation.value]
            consensus_confidence = (winning_votes / total_votes) * (
                weighted_scores[final_recommendation.value] / winning_votes
                if winning_votes > 0 else 0.5
            )
        else:
            consensus_confidence = 0.5
        
        return ConsensusResponse(
            id=0,  # Will be set when saved to DB
            symbol=symbol,
            final_recommendation=final_recommendation,
            consensus_confidence=round(consensus_confidence, 2),
            vote_breakdown=vote_breakdown,
            agent_analyses=valid_analyses,
            reasoning_summary=reasoning,
            created_at=datetime.utcnow()
        )
    
    def _generate_reasoning_summary(
        self,
        analyses: List[AgentAnalysisResult],
        vote_breakdown: Dict[str, int]
    ) -> str:
        """Generate a summary of the consensus reasoning."""
        total = sum(vote_breakdown.values())
        
        summary_parts = [
            f"Consensus reached with {total} agents participating.",
            f"Vote breakdown: {vote_breakdown['buy']} buy, {vote_breakdown['sell']} sell, {vote_breakdown['hold']} hold.",
            "",
            "Key insights from agents:"
        ]
        
        for analysis in analyses:
            summary_parts.append(
                f"- {analysis.agent_type.value.title()}: {analysis.recommendation.value.upper()} "
                f"(confidence: {analysis.confidence:.0%})"
            )
        
        return "\n".join(summary_parts)


# Global consensus engine instance
consensus_engine = ConsensusEngine()


async def analyze_symbol(
    symbol: str,
    market_data: Dict[str, Any],
    enabled_agents: List[AgentType] = None
) -> ConsensusResponse:
    """
    Analyze a symbol using the multi-agent consensus system.
    
    Args:
        symbol: Stock symbol to analyze
        market_data: Market data for analysis
        enabled_agents: List of agents to use
    
    Returns:
        ConsensusResponse with recommendation
    """
    return await consensus_engine.get_consensus(symbol, market_data, enabled_agents)


async def get_single_agent_analysis(
    agent_type: AgentType,
    symbol: str,
    market_data: Dict[str, Any]
) -> AgentAnalysisResult:
    """
    Get analysis from a single agent.
    
    Args:
        agent_type: Type of agent to use
        symbol: Stock symbol to analyze
        market_data: Market data for analysis
    
    Returns:
        AgentAnalysisResult with recommendation
    """
    agent = consensus_engine.agents.get(agent_type)
    if not agent:
        raise ValueError(f"Unknown agent type: {agent_type}")
    
    return await agent.analyze(symbol, market_data)

File: server/services/aiAgents.ts
import { invokeLLM } from "../_core/llm";
import { callDataApi } from "../_core/dataApi";

// Agent types for the 7-agent consensus system
export type AgentType = 
  | "technical" 
  | "fundamental" 
  | "sentiment" 
  | "risk" 
  | "microstructure" 
  | "macro" 
  | "quant";

export type TradingSignal = "strong_buy" | "buy" | "hold" | "sell" | "strong_sell";

export interface AgentAnalysisResult {
  agent: AgentType;
  score: number; // -1 to 1
  signal: TradingSignal;
  confidence: number; // 0 to 1
  reasoning: string;
  keyFactors: string[];
}

export interface ConsensusResult {
  symbol: string;
  timestamp: Date;
  agents: AgentAnalysisResult[];
  consensusScore: number;
  consensusSignal: TradingSignal;
  overallConfidence: number;
  recommendation: string;
  riskAssessment: string;
}

// Fetch market data from Yahoo Finance
export async function getMarketData(symbol: string) {
  try {
    const response = await callDataApi("YahooFinance/get_stock_chart", {
      query: {
        symbol,
        region: "US",
        interval: "1d",
        range: "3mo",
        includeAdjustedClose: true,
      },
    });
    return response;
  } catch (error) {
    console.error(`Failed to fetch market data for ${symbol}:`, error);
    return null;
  }
}

// Fetch stock insights
export async function getStockInsights(symbol: string) {
  try {
    const response = await callDataApi("YahooFinance/get_stock_insights", {
      query: { symbol },
    });
    return response;
  } catch (error) {
    console.error(`Failed to fetch insights for ${symbol}:`, error);
    return null;
  }
}

// Fetch stock holders data
export async function getStockHolders(symbol: string) {
  try {
    const response = await callDataApi("YahooFinance/get_stock_holders", {
      query: { symbol, region: "US", lang: "en-US" },
    });
    return response;
  } catch (error) {
    console.error(`Failed to fetch holders for ${symbol}:`, error);
    return null;
  }
}

// Convert score to trading signal
function scoreToSignal(score: number): TradingSignal {
  if (score >= 0.6) return "strong_buy";
  if (score >= 0.2) return "buy";
  if (score >= -0.2) return "hold";
  if (score >= -0.6) return "sell";
  return "strong_sell";
}

// Technical Analysis Agent
async function runTechnicalAgent(symbol: string, marketData: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Technical Analysis AI Agent. Analyze the following market data for ${symbol} and provide a trading signal.

Market Data Summary:
${JSON.stringify(marketData?.chart?.result?.[0]?.meta || {}, null, 2)}

Recent price action and volume data is available. Consider:
- Price trends (moving averages, support/resistance)
- Volume patterns
- Momentum indicators (RSI, MACD concepts)
- Chart patterns

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a technical analysis expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "technical_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "technical",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Technical analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Technical agent error:", error);
    return {
      agent: "technical",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete technical analysis",
      keyFactors: []
    };
  }
}

// Fundamental Analysis Agent
async function runFundamentalAgent(symbol: string, insights: any, holders: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Fundamental Analysis AI Agent. Analyze the following data for ${symbol}.

Stock Insights:
${JSON.stringify(insights || {}, null, 2).slice(0, 2000)}

Institutional Holders Summary:
${JSON.stringify(holders?.quoteSummary?.result?.[0]?.institutionOwnership || {}, null, 2).slice(0, 1000)}

Consider:
- Valuation metrics (P/E, P/B, PEG)
- Growth prospects
- Competitive position
- Institutional ownership changes
- Earnings quality

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a fundamental analysis expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "fundamental_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "fundamental",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Fundamental analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Fundamental agent error:", error);
    return {
      agent: "fundamental",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete fundamental analysis",
      keyFactors: []
    };
  }
}

// Sentiment Analysis Agent
async function runSentimentAgent(symbol: string, insights: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Sentiment Analysis AI Agent. Analyze market sentiment for ${symbol}.

Available Insights:
${JSON.stringify(insights || {}, null, 2).slice(0, 2000)}

Consider:
- News sentiment
- Social media trends
- Analyst ratings
- Insider trading activity
- Options market sentiment

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a sentiment analysis expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "sentiment_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "sentiment",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Sentiment analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Sentiment agent error:", error);
    return {
      agent: "sentiment",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete sentiment analysis",
      keyFactors: []
    };
  }
}

// Risk Management Agent
async function runRiskAgent(symbol: string, marketData: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Risk Management AI Agent. Assess the risk profile for trading ${symbol}.

Market Data:
${JSON.stringify(marketData?.chart?.result?.[0]?.meta || {}, null, 2)}

Consider:
- Volatility levels
- Liquidity risk
- Correlation with market
- Maximum drawdown potential
- Position sizing recommendations

Respond in JSON format:
{
  "score": <number between -1 and 1, where negative means high risk>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a risk management expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "risk_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "risk",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Risk analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Risk agent error:", error);
    return {
      agent: "risk",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete risk analysis",
      keyFactors: []
    };
  }
}

// Market Microstructure Agent
async function runMicrostructureAgent(symbol: string, marketData: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Market Microstructure AI Agent. Analyze order flow and market structure for ${symbol}.

Market Data:
${JSON.stringify(marketData?.chart?.result?.[0]?.meta || {}, null, 2)}

Consider:
- Bid-ask spread patterns
- Volume profile
- Order flow imbalances
- Market maker activity
- Execution quality factors

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a market microstructure expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "microstructure_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "microstructure",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Microstructure analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Microstructure agent error:", error);
    return {
      agent: "microstructure",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete microstructure analysis",
      keyFactors: []
    };
  }
}

// Macroeconomic Agent
async function runMacroAgent(symbol: string): Promise<AgentAnalysisResult> {
  const prompt = `You are a Macroeconomic Analysis AI Agent. Analyze the macroeconomic environment for trading ${symbol}.

Consider current macroeconomic factors:
- Interest rate environment
- Inflation trends
- GDP growth expectations
- Currency movements
- Sector rotation patterns
- Global economic conditions

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a macroeconomic analysis expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "macro_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "macro",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Macroeconomic analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Macro agent error:", error);
    return {
      agent: "macro",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete macroeconomic analysis",
      keyFactors: []
    };
  }
}

// Quantitative Agent
async function runQuantAgent(symbol: string, marketData: any): Promise<AgentAnalysisResult> {
  const prompt = `You are a Quantitative Analysis AI Agent. Apply quantitative models to analyze ${symbol}.

Market Data:
${JSON.stringify(marketData?.chart?.result?.[0]?.meta || {}, null, 2)}

Apply quantitative methods:
- Statistical arbitrage signals
- Mean reversion indicators
- Momentum factors
- Volatility modeling
- Machine learning pattern recognition

Respond in JSON format:
{
  "score": <number between -1 and 1>,
  "confidence": <number between 0 and 1>,
  "reasoning": "<brief explanation>",
  "keyFactors": ["<factor1>", "<factor2>", "<factor3>"]
}`;

  try {
    const response = await invokeLLM({
      messages: [
        { role: "system", content: "You are a quantitative analysis expert. Always respond with valid JSON." },
        { role: "user", content: prompt }
      ],
      response_format: {
        type: "json_schema",
        json_schema: {
          name: "quant_analysis",
          strict: true,
          schema: {
            type: "object",
            properties: {
              score: { type: "number" },
              confidence: { type: "number" },
              reasoning: { type: "string" },
              keyFactors: { type: "array", items: { type: "string" } }
            },
            required: ["score", "confidence", "reasoning", "keyFactors"],
            additionalProperties: false
          }
        }
      }
    });

    const content = response.choices[0].message.content;
    const result = JSON.parse(typeof content === 'string' ? content : "{}");
    return {
      agent: "quant",
      score: Math.max(-1, Math.min(1, result.score || 0)),
      signal: scoreToSignal(result.score || 0),
      confidence: Math.max(0, Math.min(1, result.confidence || 0.5)),
      reasoning: result.reasoning || "Quantitative analysis completed",
      keyFactors: result.keyFactors || []
    };
  } catch (error) {
    console.error("Quant agent error:", error);
    return {
      agent: "quant",
      score: 0,
      signal: "hold",
      confidence: 0.3,
      reasoning: "Unable to complete quantitative analysis",
      keyFactors: []
    };
  }
}

// Run all agents and compute consensus
export async function runAgentConsensus(
  symbol: string, 
  enabledAgents: AgentType[] = ["technical", "fundamental", "sentiment", "risk", "microstructure", "macro", "quant"]
): Promise<ConsensusResult> {
  // Fetch all required data
  const [marketData, insights, holders] = await Promise.all([
    getMarketData(symbol),
    getStockInsights(symbol),
    getStockHolders(symbol)
  ]);

  // Run enabled agents in parallel
  const agentPromises: Promise<AgentAnalysisResult>[] = [];
  
  if (enabledAgents.includes("technical")) {
    agentPromises.push(runTechnicalAgent(symbol, marketData));
  }
  if (enabledAgents.includes("fundamental")) {
    agentPromises.push(runFundamentalAgent(symbol, insights, holders));
  }
  if (enabledAgents.includes("sentiment")) {
    agentPromises.push(runSentimentAgent(symbol, insights));
  }
  if (enabledAgents.includes("risk")) {
    agentPromises.push(runRiskAgent(symbol, marketData));
  }
  if (enabledAgents.includes("microstructure")) {
    agentPromises.push(runMicrostructureAgent(symbol, marketData));
  }
  if (enabledAgents.includes("macro")) {
    agentPromises.push(runMacroAgent(symbol));
  }
  if (enabledAgents.includes("quant")) {
    agentPromises.push(runQuantAgent(symbol, marketData));
  }

  const agentResults = await Promise.all(agentPromises);

  // Calculate weighted consensus
  let totalWeight = 0;
  let weightedScore = 0;
  let totalConfidence = 0;

  // Agent weights (can be customized)
  const agentWeights: Record<AgentType, number> = {
    technical: 1.0,
    fundamental: 1.0,
    sentiment: 0.8,
    risk: 1.2, // Risk has higher weight
    microstructure: 0.7,
    macro: 0.8,
    quant: 1.0
  };

  for (const result of agentResults) {
    const weight = agentWeights[result.agent] * result.confidence;
    weightedScore += result.score * weight;
    totalWeight += weight;
    totalConfidence += result.confidence;
  }

  const consensusScore = totalWeight > 0 ? weightedScore / totalWeight : 0;
  const overallConfidence = agentResults.length > 0 ? totalConfidence / agentResults.length : 0;

  // Generate recommendation
  const consensusSignal = scoreToSignal(consensusScore);
  let recommendation = "";
  let riskAssessment = "";

  const riskAgent = agentResults.find(a => a.agent === "risk");
  if (riskAgent) {
    riskAssessment = riskAgent.reasoning;
  }

  switch (consensusSignal) {
    case "strong_buy":
      recommendation = `Strong bullish consensus for ${symbol}. ${agentResults.length} agents agree on positive outlook with ${(overallConfidence * 100).toFixed(0)}% average confidence.`;
      break;
    case "buy":
      recommendation = `Moderate bullish signal for ${symbol}. Consider entering a position with appropriate risk management.`;
      break;
    case "hold":
      recommendation = `Neutral signal for ${symbol}. Wait for clearer direction before taking action.`;
      break;
    case "sell":
      recommendation = `Moderate bearish signal for ${symbol}. Consider reducing exposure or taking profits.`;
      break;
    case "strong_sell":
      recommendation = `Strong bearish consensus for ${symbol}. Consider exiting positions or implementing hedges.`;
      break;
  }

  return {
    symbol,
    timestamp: new Date(),
    agents: agentResults,
    consensusScore,
    consensusSignal,
    overallConfidence,
    recommendation,
    riskAssessment
  };
}

// Get available agents based on subscription tier
export function getAvailableAgents(tier: string): AgentType[] {
  switch (tier) {
    case "free":
      return ["technical", "sentiment"];
    case "starter":
      return ["technical", "fundamental", "sentiment", "risk"];
    case "pro":
    case "elite":
      return ["technical", "fundamental", "sentiment", "risk", "microstructure", "macro", "quant"];
    default:
      return ["technical", "sentiment"];
  }
}
